platform/qemu/virt/platform.c

```diff
@@ -122,8 +127,29 @@ static int virt_timer_init(bool cold_boot)
 
 static int virt_system_down(u32 type)
 {
-	/* For now nothing to do. */
-	return 0;
+	struct sbi_scratch *local_scratch = NULL;
+	int ret = 0;
+
+	/* Tell the "finisher" that the simulation
+	 * was successful so that QEMU exits
+	 */
+	writew(VIRT_TEST_FINISHER_PASS, (void *)VIRT_TEST_ADDR);
+
+	/* The above didn't work, send an IPI on
+	 * everyone to hang
+	 */
+	local_scratch = sbi_scratch_thishart_ptr();
+
+	ret = sbi_ipi_send_many(local_scratch,
+				NULL, SBI_IPI_EVENT_HALT);
+
+	/* We should be done by now, if not something
+	 * weird happened, all we can do is hang
+	 */
+	sbi_hart_hang();
+
```

Comment:
We already have sbi_hart_hang() in sbi_system_reboot() and sbi_system_shutdown() so this is not required here.

Reply:
I get it for sbi_system_shutdown but why on reboot ?

Reply:
In both shutdown and reboot, if platform specific method fails to do the job then we should atleast halt other HARTs. Do you agree ?

Reply:
I'm ok if we halt all other harts (vs everyone) since the current hart will then be able to do the reboot, but halting everyone or everyone else and then the current hart doesn't make sense for reboot (this way reboot == halt). We can simply return an error and let the OS handle it (e.g. on Linux we could trigger an architecture-independent restart method, check out https://elixir.bootlin.com/linux/latest/source/include/linux/reboot.h#L64).

Also reboot != reset, we need to have a platform-specific reset method where all other harts are hanged and the current one triggers a reset through a power domain controller/pmic/gpio pin/whatever and if that fails it'll fall-back to reboot. To me reboot means "clean up and execute the second stage boot loader" (e.g. u-boot), or in case of the kernel "clean up and re-execute the kernel". Reset on the other hand is "power-cycle the system".

That means in case of re-boot we shouldn't use sbi_hart_hang on the other harts because since the while loop is there it will never recover. I think we should have an sbi_hart_wait function where we disable external interrupts (so that another hart can't get an interrupt from a device and try to execute code driver code during reboot) and mark the hart unavailable (so if the OS sends an IPI during reboot it'll get ignored), which basically means wfi where interrupt == IPI (IPI will get un-blocked after re-boot where all harts will be marked as available again).

What do you think ?

Reply:
I think we are on same page here.

Currently, we don't have a clear definition of Reboot SBI call in SBI spec so our reboot implementation is crude.

I think we should wait for SBI PM extension to do changes in sbi_system_reboot() implementation so let's leave it unmodified now.